cip/1.accepted/CIP2016-12-14-Constraint-syntax.adoc

```diff
@@ -0,0 +1,285 @@
+
+== Proposal
+
+This CIP includes the following proposed constraints:
+
+* <<uniqueness>>
+* <<existence, Node property existence constraint>>
+* <<existence, Relationship property existence constraint>>
+
+Each constraint is detailed in its own below section.
+
+Once a constraint has been created, it may not be amended.
+Should a user wish to change its definition, it has to be dropped and recreated with an updated structure.
+
+==== Constraint names
+
+All constraints require the user to specify a nonempty _name_ at constraint creation time.
+This name is subsequently the handle with which a user may refer to the constraint, e.g. when dropping it.
+
+// TODO: Should we impose restrictions on the domain of constraint names, or are all Unicode characters allowed?
+
+=== Syntax overview
+
+The syntax for all constraints follow the same basic outline.
+
+.Grammar definition for constraint syntax.
+[source, ebnf]
+----
+constraint command = create-constraint | drop-constraint ;
+create-constraint  = "CREATE", "CONSTRAINT", constraint-name, "FOR", constraint-pattern, "REQUIRE", constraint-expr ;
+constraint-name    = symbolic-name
+constraint-pattern = node-pattern | simple-pattern ;
+constraint-expr    = uniqueness-constraint | existence-constraint ;
+drop-constraint    = "DROP", "CONSTRAINT", constraint-name ;
+----
+
+The constraint expressions vary depending on the actual constraint (see the detailed sections).
+
+.Example of dropping a constraint with name foo:
+[source, cypher]
+----
+DROP CONSTRAINT foo
+----
+
+=== Semantics overview
+
+The semantics are defined for each type of constraint, but some characteristics are shared:
+
+* When a statement tries to create a constraint on a graph where the data does not pass the constraint criterion, that statement will raise an error.
+* When a statement tries to create a constraint with a name that already exists, that statement will raise an error.
+* When a statement tries to drop a constraint referencing a name that does not exist, that statement will raise an error.
+* When an updating statement tries to modify the graph in such a way that it would violate a constraint, that statement will raise an error.
+
+The constraints define a _domain_ within which the constraint applies.
+The domain is defined by the constraint pattern.
+
+[[uniqueness]]
+=== Node property uniqueness constraint
+
+This constraint enforces that there can not be duplicate values of a certain property for a certain type of node.
+For example, that among nodes labeled with `:Person`, each `email` property must be unique.
+
+==== Syntax
+
+.Grammar definition for node property uniqueness constraint:
+[source, ebnf]
+----
+uniqueness-constraint = "UNIQUE", property-expression, { ",", property-expression } ;
+----
+
+.Example of single-property uniqueness constraint:
+[source, cypher]
+----
+CREATE CONSTRAINT unique_person_email
```

Comment:
Are we notionally mandating this particular style (for a constraint name)? Ie. lower case & underscores?

Reply:
No, it is just an example. It may be any symbolic string.